src/WebJobs.Script/Binding/Manual/ManualTriggerAttributeBindingProvider.cs

```diff
@@ -0,0 +1,146 @@
+            {
+                _parameter = parameter;
+                _isUserTypeBinding = isUserTypeBinding;
+
+                if (_isUserTypeBinding)
+                {
+                    // Create the BindingDataProvider from the user Type. The BindingDataProvider
+                    // is used to define the binding parameters that the binding exposes to other
+                    // bindings (i.e. the properties of the POCO can be bound to by other bindings).
+                    // It is also used to extract the binding data from an instance of the Type.
+                    _bindingDataProvider = BindingDataProvider.FromType(parameter.ParameterType);
+                }
+            }
+
+            public IReadOnlyDictionary<string, Type> BindingDataContract
+            {
+                get
+                {
+                    // if we're binding to a user Type, we'll have a contract,
+                    // otherwise none
+                    return _bindingDataProvider?.Contract;
+                }
+            }
+
+            public Type TriggerValueType
+            {
+                get { return _parameter.ParameterType; }
+            }
+
+            public Task<ITriggerData> BindAsync(object value, ValueBindingContext context)
+            {
+                IValueProvider valueProvider = null;
+                IReadOnlyDictionary<string, object> bindingData = null;
+                string invokeString = value?.ToString();
+                if (_isUserTypeBinding)
+                {
+                    if (value != null && value.GetType() != _parameter.ParameterType)
+                    {
+                        if (value is string)
+                        {
+                            value = JsonConvert.DeserializeObject((string)value, _parameter.ParameterType);
+                        }
+                        else
+                        {
+                            throw new InvalidOperationException($"Unable to bind value to type {_parameter.ParameterType}");
+                        }
+                    }
+
+                    valueProvider = new SimpleValueProvider(_parameter.ParameterType, value, invokeString);
+                    if (_bindingDataProvider != null)
+                    {
+                        // binding data is defined by the user type
+                        // the provider might be null if the Type is invalid, or if the Type
+                        // has no public properties to bind to
+                        bindingData = _bindingDataProvider.GetBindingData(valueProvider.GetValue());
+                    }
+                }
+                else
+                {
+                    valueProvider = new SimpleValueProvider(_parameter.ParameterType, value, invokeString);
+                    var bindingDataTemp = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+                    Utility.ApplyBindingData((string)value, bindingDataTemp);
```

Comment:
This cast will fail if value is null. Since `ApplyBindingData` takes an object, we shouldn't need this cast.